### PR TITLE
feat(SSA): simplify array_get on slice_insert

### LINKED ISSUE #10278: `as_slice` could be simplified in more cases
Right now `as_slice` is only simplified during SSA when its argument is a known array. For example, in this Noir program:

```noir
fn main() -> pub Field {
    let array = [1, 2, 3];
    let slice = array.as_slice();
    let slice = slice.insert(0, 1);
    slice[0]
}
```

the call to `as_slice` doesn't show up in any SSA stage because it's immediately simplified to `make_array [1, 2, 3]`.

This isn't the case when the array is not known. For example, in this Noir program:

```noir
fn main(array: [Field; 3]) -> pub Field {
    let slice = array.as_slice();
    let slice = slice.insert(0, 1);
    slice[0]
}
```

we get this final SSA:

```
acir(inline) predicate_pure fn main f0 {
  b0(v0: [Field; 3]):
    v2, v3 = call as_slice(v0) -> (u32, [Field])
    v8, v9 = call slice_insert(u32 3, v3, u32 0, Field 1) -> (u32, [Field])
    constrain v8 != u32 0, "Index out of bounds"
    v10 = array_get v9, index u32 0 -> Field
    return v10
}
```

Note how there's an "index out of bounds" check, the call to `as_slice` is still there, etc., However, the compiler should be able to figure out that the slice came from the array, that it now has a length of 4, and even be able to say the output of the program is 1.

A potential optimization is to replace the `as_slice` call with a slice where each element corresponds to the array elements. We can do this manually in Noir:

```noir
fn main(array: [Field; 3]) -> pub Field {
    let slice = &[array[0], array[1], array[2]];
    let slice = slice.insert(0, 1);
    slice[0]
}
```

Now if we compile and see the final SSA, it's this:

```
acir(inline) pure fn main f0 {
  b0(v0: [Field; 3]):
    return Field 1
}
```

The compiler should then apply this manual optimization.

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem

Resolves #10278

## Summary

Alternative to #10279 with what's written in https://github.com/noir-lang/noir/pull/10279#issuecomment-3451174767

## Documentation

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/opt/slice_intrinsics_length.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+                Some((original_slice_length, length.into()))
+            } else if slice_insert.is_some_and(|op| target_func == &op) {
+                if let Some(length) = context.dfg.get_numeric_constant(arguments[0]) {
+                    // For `slice_insert(length, ...)` we can replace the resulting length with length + 1
+                    let length = length + FieldElement::one();
+                    let new_slice_length = context.dfg.instruction_results(instruction_id)[0];
+                    Some((new_slice_length, length))
+                } else {
+                    None
+                }
```
Comment (vezenovm) [inline compiler/noirc_evaluator/src/ssa/opt/slice_intrinsics_length.rs:86]:
```suggestion
                // For `slice_insert(length, ...)` we can replace the resulting length with length + 1
                context.dfg.get_numeric_constant(arguments[0]).map(|length| {
                    let length = length + FieldElement::one();
                    let new_slice_length = context.dfg.instruction_results(instruction_id)[0];
                    (new_slice_length, length)
                })
```
nit: option for reducing nesting

File: compiler/noirc_evaluator/src/ssa/opt/slice_intrinsics_length.rs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            } else if slice_remove.is_some_and(|op| target_func == &op) {
+                if let Some(length) = context.dfg.get_numeric_constant(arguments[0]) {
```
Comment (vezenovm) [inline compiler/noirc_evaluator/src/ssa/opt/slice_intrinsics_length.rs:90]: same as https://github.com/noir-lang/noir/pull/10295/files#r2466487398

File: compiler/noirc_evaluator/src/ssa/opt/slice_intrinsics_length.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    /// A simple SSA pass to find any calls to slice intrinsics  and replacing any references to the length of the
+    /// resulting slice with the length of the array from which it was generated, or with a relative
+    /// length based on the input length.
+    ///
+    /// This allows the length of a slice generated from an array to be used in locations where a constant value is
+    /// necessary when the value of the array is unknown.
+    ///
+    /// Note that this pass must be placed before loop unrolling to be useful.
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn slice_instrinsics_length_optimization(mut self) -> Self {
```
Comment (vezenovm) [inline compiler/noirc_evaluator/src/ssa/opt/slice_intrinsics_length.rs:22]:
```suggestion
    pub(crate) fn slice_intrinsics_length_optimization(mut self) -> Self {
```

File: compiler/noirc_evaluator/src/ssa/ir/dfg/simplify.rs [hunk 4]
```diff
@@ -426,6 +449,76 @@ fn try_optimize_array_get_from_previous_set(
...
+
+    let Instruction::Call { func, arguments } = &dfg[instruction] else {
+        return SimplifyResult::None;
+    };
+
+    if func != slice_insert {
+        return SimplifyResult::None;
+    }
+
+    // slice_insert(length, slice, index, values...)
```
Comment (vezenovm) [inline compiler/noirc_evaluator/src/ssa/ir/dfg/simplify.rs:478]: Do we need this still?

File: compiler/noirc_evaluator/src/ssa/opt/slice_intrinsics_length.rs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        ");
+    }
+
+    #[test]
+    fn slice_insert_optimization() {
+        let src = "
+        acir(inline) fn main f0 {
+          b0():
+            v0 = make_array [Field 2, Field 3] : [Field]
+            v1, v2 = call slice_insert(u32 2, v0, u32 1, Field 4) -> (u32, [Field])
```
Comment (vezenovm) [inline compiler/noirc_evaluator/src/ssa/opt/slice_intrinsics_length.rs:189]: Would be good to add a test for the negative case that when we have a dynamic value for the length but a constant array we do not do anything.

**Comment by asterite** [comment]
> I do think these optimizations make sense. They are not super invasive so I would be ok with adding them (especially the parameter array change in try_optimize_array_get_from_previous_set).

Cool! I just read your comment, I kept pushing changes with more "slice intrinsics length" optimizations.

If it's fine with you, I think I'd prefer to split this PR into smaller PR, each with one of the proposed optimizations, just so that if there's an issue with just one of them it's easier to revert, etc.

Now I'm thinking optimizing `array_get` on `slice_insert` is good, but we can also optimize `array_get` on `slice_remove`, `slice_push_back`, etc., as long as we know the slice length at compile-time. Maybe this is just another variant not considered in `try_optimize_array_set_from_previous_get` (though the index needs to be adjusted accordingly).

I might also introduce a new SSA pass instead of changing `as_slice_length` because we recently audited it so I'd feel bad modifying it 😅 (though the pass is pretty simple...)

**Comment by vezenovm** [comment]
> If it's fine with you, I think I'd prefer to split this PR into smaller PR, each with one of the proposed optimizations, just so that if there's an issue with just one of them it's easier to revert, etc.

Yeah I think that would be a good idea.

> I might also introduce a new SSA pass instead of changing `as_slice_length` because we recently audited it so I'd feel bad modifying it 😅 (though the pass is pretty simple...)

Whatever you think is best here. As you said the actual `as_slice` opt logic is very minimal, but it does increase our audit scope so perhaps we should just contain it all in a single pass. Especially if we are able to later remove the `as_slice_length` opt if we remove the as_slice intrinsic as per https://github.com/noir-lang/noir/pull/10279#issuecomment-3452199789.

**Comment by asterite** [comment]
Closing in favor of https://github.com/noir-lang/noir/pull/10300 and other upcoming PRs.
